### PR TITLE
feat: add mkcd function

### PR DESCRIPTION
**Pull request by DeflateAwning**
## TODOs:

- [ ] **N/A** - If addressing an issue, a commit message mentions `Fixes issue #<issue-number>`
- [ ] **Please advise if relevant** - Changes to fish usage are reflected in user documentation/manpages.
- [ ] **Please advise if required** - Tests have been added for regressions fixed
- [x] User-visible changes noted in CHANGELOG.rst

### CODE REVIEW

**Review by Vishrut2403 — COMMENTED** [review]
Might be worth handling the case where multiple arguments are passed. ```cd $argv``` will fail even if ```mkdir -p``` is used.

**Comment by DeflateAwning** [comment]
@Vishrut2403 very fair comment. Just updated to have a test/check to ensure only a single argument is passed. An alternative I considered was `cd`ing into only the last argument, but I think this is more aligned with what people would expect.

**Review by Vishrut2403 — COMMENTED** [review]
LGTM, maybe maintainers can weigh in on how they want this function implemented

**Review by Vishrut2403 — COMMENTED** [review]
A few tests for the edge cases would be much better, if this gets modified later

**Comment by DeflateAwning** [comment]
@Vishrut2403 This one is so extremely simple in scope that I don't really think a test provides any value. Can you point to a test for something similar in scope?

**Comment by Vishrut2403** [comment]
> @Vishrut2403 This one is so extremely simple in scope that I don't really think a test provides any value. Can you point to a test for something similar in scope?

My bad, I couldn't find any tests for simple functions either

File: share/functions/mkcd.fish [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+# These are very common and useful
+#
+
+function mkcd --description "Create a directory and enter it"
+    if test (count $argv) -ne 1
+        echo "mkcd: expected 1 argument, got "(count $argv) >&2
+        return 1
+    end
+    mkdir -p $argv
+    and cd $argv
```
Comment (krobelus) [inline share/functions/mkcd.fish:12]: this seems reasonable. A possible alternative would be an "abbreviation" that expands to

```
mkdir foo &&
cd foo
```

That one would add autosuggestions for `cd foo`, for better or worse

File: share/functions/mkcd.fish [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+#
+# These are very common and useful
+#
+
+function mkcd --description "Create a directory and enter it"
+    if test (count $argv) -ne 1
+        echo "mkcd: expected 1 argument, got "(count $argv) >&2
+        return 1
+    end
+    mkdir -p $argv
```
Comment (krobelus) [inline share/functions/mkcd.fish:11]: the `-p` feels opinionated, is this what other implementations do?
I feel like `-p` shouldn't be needed too often so I'm not sure if it should be the default.
We could take it as optional argument to `mkcd`.

File: share/functions/mkcd.fish (hunk 1 above)
Comment (DeflateAwning) [reply share/functions/mkcd.fish:12]: I'm not sure I follow the benefit. Feel free to explain more if you're seeing something I'm not. 

Surprisingly many ways to implement such an extremely-small 2-step command. Open to alternatives, but this is the one I deemed best.

File: share/functions/mkcd.fish (hunk 2 above)
Comment (DeflateAwning) [reply share/functions/mkcd.fish:11]: Indeed opinionated. `-p` appears to be the standard across the vast majority of `mkcd` reference implementations I looked at, yes. Not too common, but sometimes helpful. I just checked my `history | grep mkcd`, and there's a handful of cases where I've used it. 

Consider `mkcd .ssh/my_keys/` or `mkcd .github/workflows/` or `mkcd /tmp/thing_i_am_doing_right_now/v1/`. 

`mkdir` is a tool intended to be used all over scripts from the past 40 years. Makes sense it takes the "safe" approach of asserting that the parent exists. This `mkcd` is intended to be a user-facing tool added to an advanced shell, where that shell already gives the user in-line feedback of whether folders exist as they're being typed (via color feedback). The safety goals and guarantees of `mkdir` are different than the goals of this `mkcd` alias designed for speed.

File: share/functions/mkcd.fish (hunk 1 above)
Comment (krobelus) [reply share/functions/mkcd.fish:12]: actually I usually type `mkdir foo` `Enter` `cd ` `alt-.` `Enter`.
But I guess in terms of discoverability it's neither better nor worse than `mkcd`, so it seems okay to add `mkcd`.

Comment (DeflateAwning) [reply share/functions/mkcd.fish:12]: Woah that's magic. Actually lifechanging. Thanks for the tip! 

I still stand by `mkcd` being awesome, but definitely 10% less important now that I know about `Alt+Period`.

**Comment by zanchey** [comment]
yeah, I haven't been keen to add this because it seems fairly simple but there's not an clear interface that everyone would expect

https://dbohdan.com/mkcd is interesting as data point, only ~17% of mkdir followed by cd

**Comment by DeflateAwning** [comment]
I have workflow optimizations for stuff that I do a heck of a lot less than 17% of the time. 

I believe the logic in that "study" (the one that gets to 17%) is a little flawed though. Consider the following workflows:

```bash
# Option A: Best when mkcd doesn't exist. 
mkdir .ssh
vim .ssh/config

# Option B: A strange 3-step action that no sensible user would do, but is what the 17% study measured.
mkdir .ssh
cd .ssh
vim config

# Option C: Best when mkcd does exist.
mkcd .ssh
vim config
```

Option C is the fewest keystrokes, of course. Option B is a lot of extra keystrokes and an extra command (compare to Option A). Still though, users do Option B 17% of the time, despite it not even being the most efficient way to get the job done, because it simplifies the thought process to switch into the directory you just made.

**Comment by krobelus** [comment]
maybe after `mkdir foo`, we should autosuggest `cd foo`? More discoverable than `mkcd`.
That feature could be extended to `git clone ..` and similar commands. There's an issue about that somewhere.
We could prioritize recently-created directories.

**Comment by DeflateAwning** [comment]
> maybe after `mkdir foo`, we should autosuggest `cd foo`? More discoverable than `mkcd`. That feature could be extended to `git clone ..` and similar commands. There's an issue about that somewhere. We could prioritize recently-created directories.

Very reasonable! My question then is "why not both"?
